Mathlib/RingTheory/Finiteness.lean

```diff
@@ -417,6 +417,28 @@ theorem fg_iff_compact (s : Submodule R M) : s.FG ↔ CompleteLattice.IsCompactE
       exact ⟨t, ssup⟩
 #align submodule.fg_iff_compact Submodule.fg_iff_compact
 
+open TensorProduct LinearMap in
+/-- Every `x : I ⊗ M` is the image of some `y : J ⊗ M` where `J ≤ I` is finitely generated,
+under the tensor product of `J.inclusion ‹J ≤ I› : J → I` and the identity `M → M`. -/
+theorem exists_fg_le_eq_rTensor_inclusion {R M N : Type*} [CommRing R] [AddCommGroup M]
+    [AddCommGroup N] [Module R M] [Module R N] {I : Submodule R N} (x : I ⊗ M) :
+      ∃ (J : Submodule R N) (_ : J.FG) (hle : J ≤ I) (y : J ⊗ M),
+        x = rTensor M (J.inclusion hle) y := by
+  induction x using TensorProduct.induction_on with
+  | zero => refine ⟨⊥, fg_bot, zero_le _, 0, rfl⟩
+  | tmul i m =>
+    refine ⟨R ∙ i.val, ?_, ?_, ⟨i.val, mem_span_singleton_self _⟩ ⊗ₜ[R] m, rfl⟩
+    · exact fg_span_singleton i.val
+    · rewrite [span_le, Set.singleton_subset_iff]; exact Subtype.mem i
+  | add x₁ x₂ ihx₁ ihx₂ =>
+    have ⟨J₁, hfg₁, hle₁, y₁, hy₁⟩ := ihx₁
+    have ⟨J₂, hfg₂, hle₂, y₂, hy₂⟩ := ihx₂
+    refine ⟨J₁ ⊔ J₂, FG.sup hfg₁ hfg₂, sup_le hle₁ hle₂,
+      rTensor M (J₁.inclusion (le_sup_left : J₁ ≤ J₁ ⊔ J₂)) y₁ +
+        rTensor M (J₂.inclusion (le_sup_right : J₂ ≤ J₁ ⊔ J₂)) y₂, ?_⟩
+    rewrite [map_add, ← comp_apply, ← rTensor_comp, ← comp_apply, ← rTensor_comp, hy₁, hy₂]
+    rfl
```

Comment:
similarly here
```suggestion
      ∃ J : Submodule R N, J.FG ∧ ∃ (hle : J ≤ I) (y : J ⊗ M),
        x = rTensor M (J.inclusion hle) y := by
  induction x using TensorProduct.induction_on with
  | zero => exact ⟨⊥, fg_bot, zero_le _, 0, rfl⟩
  | tmul i m => exact ⟨R ∙ i.val, fg_span_singleton i.val,
      (span_singleton_le_iff_mem _ _).mpr i.prop,
      ⟨i.val, mem_span_singleton_self _⟩ ⊗ₜ[R] m, rfl⟩
  | add x₁ x₂ ihx₁ ihx₂ =>
    obtain ⟨J₁, hfg₁, hle₁, y₁, rfl⟩ := ihx₁
    obtain ⟨J₂, hfg₂, hle₂, y₂, rfl⟩ := ihx₂
    refine ⟨J₁ ⊔ J₂, hfg₁.sup hfg₂, sup_le hle₁ hle₂,
      rTensor M (J₁.inclusion (le_sup_left : J₁ ≤ J₁ ⊔ J₂)) y₁ +
        rTensor M (J₂.inclusion (le_sup_right : J₂ ≤ J₁ ⊔ J₂)) y₂, ?_⟩
    simp_rw [map_add, ← rTensor_comp_apply]
    rfl
```

Reply:
Great! I somewhat prefer the original statement with the four hypotheses in the existential. That style seems not uncommon. Slightly more strongly I prefer not to use `simp_rw` when `rewrite` suffices, since `rewrite` (or `rw`) is a lot faster. I like the `obtain` and `rTensor_comp_apply`.

Reply:
I also like the replacement of `refine` by `exact`, but `Subtype.prop` is a rather subtle theorem, and I prefer the simple accessor `Subtype.property` when the subtlety is not needed. It's easier to understand what the code is doing.

Reply:
Applied with some modifications.

Reply:
`Subtype.property` is syntactically `p x.val` while [Subtype.prop](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Subtype.html#Subtype.prop) is `p ↑x`. They used to be different in Lean 3 but are practically the same in Lean 4 due to how coercion now works. Therefore I think you can think of `Subtype.prop` as a shorthand for `Subtype.property`. (Actually I think `x.2` is also common in mathlib (`.2` for `Iff.mpr` is also common).